src/main/scala/linting/rule/LintConflictingModuleNames.scala

```diff
@@ -0,0 +1,318 @@
+  * [[NamingStrategyAnnotation]] then the transform will attempt to rename
+  * modules according to that naming strategy. The default naming strategy is
+  * [[ExactNamingStrategy]] which will throw an exception if there are any
+  * module name conflicts.
+  */
+class RenameDesiredNames extends Transform with DependencyAPIMigration {
+
+  override def prerequisites = Seq(Dependency[DedupModules])
+  override def optionalPrerequisiteOf = Seq(Dependency[VerilogEmitter])
+
+  override def invalidates(transform: Transform) = false
+
+  private def checkStrategy(
+    strategy: NamingStrategy,
+    desiredName: String,
+    modules: Seq[Module]): Option[Map[String, String]] = {
+    val fn = strategy.getName(desiredName)(_)
+    val nameMap = modules.map(m => m.name -> fn(m)).toMap
+    if (nameMap.values.toSet.size == modules.size) Some(nameMap) else None
+  }
+
+  def execute(state: CircuitState): CircuitState = {
+    val violations = new Violations()
+
+    val modMap = state.circuit.modules.collect {
+      case m: Module => m.name -> m
+    }.toMap
+
+    val namingStrategyAnnos = state.annotations.collect {
+      case a: NamingStrategyAnnotation if a.target.circuit == state.circuit.main => a
+    }
+
+    val strategyMap = namingStrategyAnnos.groupBy { a =>
+      val referringModule = Target.referringModule(a.target).module
+      require(modMap.contains(referringModule), "NamingStrategyAnnotation may not refer to blackboxes")
+      referringModule
+    }.map { case (module, annos) =>
+      val strategies = annos.map(_.strategy).distinct
+      require(strategies.size == 1, s"Conflicting naming strategies for module $module: ${strategies.mkString(", ")}")
+      module -> strategies.head
+    }
+
+    val overrideDesiredNameAnnos = state.annotations.collect {
+      case a: OverrideDesiredNameAnnotation if a.target.circuit == state.circuit.main => a
+    }
+
+    val moduleToDesiredName: mutable.Map[String, mutable.Set[String]] = mutable.Map()
+
+    val nameMap = overrideDesiredNameAnnos.groupBy(_.desiredName).mapValues { annos =>
+      annos.map(a => Target.referringModule(a.target).module).distinct.map { referringModule =>
+        require(modMap.contains(referringModule), "ModuleNameAnnotations may not refer to blackboxes")
+        val desiredNames = moduleToDesiredName.getOrElseUpdate(referringModule, mutable.Set())
+        desiredNames += annos.head.desiredName
+        modMap(referringModule)
+      }
+    }
+
+    val conflictingDesiredNames = moduleToDesiredName.collect {
+      case kv@ (moduleName, desiredName) if desiredName.size > 1 => kv
+    }
+
+    require(conflictingDesiredNames.size == 0, {
+      val explanation = conflictingDesiredNames.map {
+        case (modName, desiredNames) => s"  ${modName}: ${desiredNames.mkString(", ")}"
+      }.mkString("\n")
+      s"Modules may not have more than one desiredName:\n${explanation}"
+    })
+
+    val renamedDesiredNames = mutable.Set[String]()
```

Comment:
Sorry, I am still confused by the interaction between `OverrideDesiredName` annotation and the `NamingStrategyAnnotation`. If I don't put a `NamingStrategyAnnotation` on an instance, but I do put a `OverrideDesiredName` annotation on it, will it be renamed to the string specified in the OverrideDesiredName annotation or not?

Reply:
yes it will be renamed